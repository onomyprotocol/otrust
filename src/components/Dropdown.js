import React, { useState } from 'react';
import styled from 'styled-components';

import { borderRadius } from 'context/responsive/cssSizes';

const DropDownContainer = styled('div')`
  width: 5em;
  margin: 0 auto;

  &:hover {
    cursor: pointer;
  }
`;

const DropDownHeader = styled('div')`
  height: 2rem;
  box-shadow: 0 2px 3px rgba(0, 0, 0, 0.15);
  font-weight: 500;
  font-size: 1.3rem;
  color: ${props => props.theme.colors.background};
  line-height: 2rem;
  text-align: center;
  vertical-align: middle;
  border-radius: ${borderRadius};
`;

const DropDownListContainer = styled('div')`
  position: absolute;
  z-index: 100;
  width: 10.5em;
`;

const DropDownList = styled('ul')`
  padding: 0;
  margin: 0;
  padding-left: 1em;
  background: #ffffff;
  border: 2px solid #e5e5e5;
  box-sizing: border-box;
  color: ${props => props.theme.colors.background};
  font-size: 1.3rem;
  font-weight: 500;
  &:first-child {
    padding-top: 0.8em;
  }
`;

const ListItem = styled('li')`
  list-style: none;
  margin-bottom: 0.8em;
  &:hover {
    color: ${props => props.theme.colors.headerBackground};
  }
`;

const options = ['ETH', 'NOM'];

export default function Dropdown({ denom, setDenom }) {
  const [isOpen, setIsOpen] = useState(false);

  const toggling = () => setIsOpen(!isOpen);

  const onOptionClicked = value => () => {
    setDenom(value);
    setIsOpen(false);
  };

  return (
    <DropDownContainer>
      <DropDownHeader onClick={toggling}>{denom || 'ETH'}</DropDownHeader>
      {isOpen && (
        <DropDownListContainer>
          <DropDownList>
            {options.map(option => (
              <ListItem onClick={onOptionClicked(option)} key={Math.random()}>
                {option}
              </ListItem>
            ))}
          </DropDownList>
        </DropDownListContainer>
      )}
    </DropDownContainer>
  );
}
