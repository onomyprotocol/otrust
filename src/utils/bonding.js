const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
