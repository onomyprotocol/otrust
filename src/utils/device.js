const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
