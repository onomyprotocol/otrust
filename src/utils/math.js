import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
