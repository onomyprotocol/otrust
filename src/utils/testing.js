import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'Theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const renderWithContext = (Component, props) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider value={{ supplyNOM: new BigNumber(0), theme: darkNew }}>
          <ExchangeContext.Provider value={{ askAmount: 0, bidAmount: 0, bidDenom: 'strong' }}>
            <Component {...props} />
          </ExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
