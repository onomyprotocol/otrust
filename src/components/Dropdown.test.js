import { cleanup, fireEvent } from '@testing-library/react';

import Dropdown from './Dropdown';
import { renderWithTheme } from '../utils/testing';

const testProps = {
  setDenom: jest.fn(),
  denom: 'DENOM',
};

describe('Given the Dropdown component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(Dropdown, testProps);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should include denom value in the header if denom was provided in the props', () => {
      const { getByText } = renderWithTheme(Dropdown, testProps);
      expect(getByText('DENOM')).toBeInTheDocument();
    });

    it('should include ETH value in the header if denom was not provided in the props', () => {
      const { getByText } = renderWithTheme(Dropdown, { ...testProps, denom: undefined });
      expect(getByText('ETH')).toBeInTheDocument();
    });

    it('should not show dropdown container', () => {
      const { queryByTestId } = renderWithTheme(Dropdown, testProps);
      expect(queryByTestId('dropdown-list-container')).toBeNull();
    });

    describe('and user clicks on dropdown header', () => {
      it('should show dropdown list container', async () => {
        const { queryByTestId } = renderWithTheme(Dropdown, testProps);
        fireEvent.click(queryByTestId('dropdown-header'));
        expect(queryByTestId('dropdown-list-container')).toBeInTheDocument();
      });

      describe('and user clicks on dropdown item', () => {
        it('should call setDenom function with option value in function params and hide dropdown-list-container', async () => {
          const { queryByTestId } = renderWithTheme(Dropdown, testProps);
          fireEvent.click(queryByTestId('dropdown-header'));
          fireEvent.click(queryByTestId('ETH'));
          expect(testProps.setDenom).toHaveBeenCalled();
          expect(testProps.setDenom).toHaveBeenCalledWith('ETH');
          expect(queryByTestId('dropdown-list-container')).toBeNull();
        });
      });
    });
  });
});
