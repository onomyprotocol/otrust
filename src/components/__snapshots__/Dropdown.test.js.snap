// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Dropdown component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM fMyXwL"
  >
    <div
      class="sc-gtsrHT jhbec"
      data-testid="dropdown-header"
    >
      DENOM
    </div>
  </div>
</DocumentFragment>
`;
